File: stackoverflow_question.py
# -*- coding: utf-8 -*-
import pywt
import numpy as np


def wrcoef(X, coef_type, coeffs, wavename, level):
    N = np.array(X).size
    a, ds = coeffs[0], list(reversed(coeffs[1:]))

    if coef_type == 'a':
        return pywt.upcoef('a', a, wavename, level=level)[:N]
    elif coef_type == 'd':
        return pywt.upcoef('d', ds[level-1], wavename, level=level)[:N]
    else:
        raise ValueError("Invalid coefficient type: {}".format(coef_type))


def decomposite(signal, coef_type='d', wname='db6', level=9):
    w = pywt.Wavelet(wname)
    a = signal
    ca = []
    cd = []
    for i in range(level):
        (a, d) = pywt.dwt(a, w)
        ca.append(a)
        cd.append(d)
    rec_a = []
    rec_d = []
    for i, coeff in enumerate(ca):
        coeff_list = [coeff, None] + [None] * i
        rec_a.append(pywt.waverec(coeff_list, w))
    for i, coeff in enumerate(cd):
        coeff_list = [None, coeff] + [None] * i
        rec_d.append(pywt.waverec(coeff_list, w))
    if coef_type == 'd':
        return rec_d
    return rec_a


if __name__ == '__main__':
    # # question
    # DATA = range(30)
    # N_LEVELS = 2
    # WAVELET_NAME = 'db4'
    # coeffs = pywt.wavedec(DATA, WAVELET_NAME, level=N_LEVELS)
    # # A2 = wrcoef(DATA, 'a', coeffs, WAVELET_NAME, N_LEVELS)
    # D2 = wrcoef(DATA, 'd', coeffs, WAVELET_NAME, N_LEVELS)
    # D1 = wrcoef(DATA, 'd', coeffs, WAVELET_NAME, 1)
    # print(D2, D1)

    # answer
    DATA = range(20)
    N_LEVELS = 4
    WAVELET_NAME = 'db1'
    coeffs = pywt.wavedec(DATA, WAVELET_NAME, level=N_LEVELS)
    # A4 = wrcoef(DATA, 'a', coeffs, WAVELET_NAME, N_LEVELS)
    D4 = wrcoef(DATA, 'd', coeffs, WAVELET_NAME, N_LEVELS)
    D3 = wrcoef(DATA, 'd', coeffs, WAVELET_NAME, 3)
    D2 = wrcoef(DATA, 'd', coeffs, WAVELET_NAME, 2)
    D1 = wrcoef(DATA, 'd', coeffs, WAVELET_NAME, 1)
    print(D4, D3, D2, D1)

    rec_d = decomposite(DATA, 'd', WAVELET_NAME, level=N_LEVELS)
    print(rec_d)
    # # Define wavelet parameters
    # wavelet_type = 'sym3'
    # wavelet_level = 3
    #
    # # Generate input data
    # N = 9
    # signal = range(N)
    #
    # # Define wavelet
    # w = pywt.Wavelet(wavelet_type)
    # # print([w.dec_lo, w.dec_hi, w.rec_lo, w.rec_hi])
    #
    # C, L = wavedec(signal, wavelet=w, level=wavelet_level)
    #
    # for n in range(len(L)-2):
    #     print(n+1)
    #     D = wrcoef(C, L, wavelet=w, level=n+1)
    #     print(D)

File: stackoverflow_question.m
%% Stackoverflow questions
% https://stackoverflow.com/questions/45051106/multilevel-partial-wavelet-reconstruction-with-pywavelets/47591523#47591523
% and
% https://dsp.stackexchange.com/questions/42358/multilevel-partial-wavelet-reconstruction-with-pywavelets/42372


%% Question 1
DATA = 0:9;
N_LEVELS = 2;
WAVELET_NAME = 'db4';
[C, L] = wavedec(DATA, N_LEVELS, WAVELET_NAME);
% A2 = wrcoef('a', C, L, WAVELET_NAME, 2)
D2 = wrcoef('d', C, L, WAVELET_NAME, 2);
D1 = wrcoef('d', C, L, WAVELET_NAME, 1);
D2, D1


%% Question 2
% level = 4
% X = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17]
% coeffs = pywt.wavedec(X, 'db1', level=level)
% A4 = wrcoef(X, 'a', coeffs, 'db1', level)
% D4 = wrcoef(X, 'd', coeffs, 'db1', level)
% D3 = wrcoef(X, 'd', coeffs, 'db1', 3)
% D2 = wrcoef(X, 'd', coeffs, 'db1', 2)
% D1 = wrcoef(X, 'd', coeffs, 'db1', 1)
% print A4 + D4 + D3 + D2 + D1
DATA = 0:17;
N_LEVELS = 4;
WAVELET_NAME = 'db1';
[C, L] = wavedec(DATA, N_LEVELS, WAVELET_NAME); 
% A4 = wrcoef('a', C, L, WAVELET_NAME, N_LEVELS);
D4 = wrcoef('d', C, L, WAVELET_NAME, N_LEVELS);
D3 = wrcoef('d', C, L, WAVELET_NAME, 3);
D2 = wrcoef('d', C, L, WAVELET_NAME, 2);
D1 = wrcoef('d', C, L, WAVELET_NAME, 1);

D4, D3, D2, D1